sycl/doc/design/VirtualFunctions.md

```diff
@@ -0,0 +1,428 @@
+
+However, we do need to filter out those virtual functions which are not
+considered to be device  as defined by the [extension specification][1], such
+as:
+
+- virtual member functions annotated with `indirectly_callable` compile-time
+  property should be emitted into device code;
+- virtual member function *not* annotated with `indirectly_callable`
+  compile-time property should *not* be emitted into device code;
+
+There is no need to actually check which exact property is applied to a
+function, it is enough to check if `add_ir_attribute_function` attribute was
+applied and that we are in SYCL device mode to decide whether or not a virtual
+function should be emitted into vtable and device code.
+
+**TODO:** any extra diagnostics we would like to emit? Like kernel without
+`calls_indirectly` property performing virtual function call.
+
+### Changes to the compiler middle-end
+
+#### Aspects propagation
+
+Aspects propagation pass should be extended to not only gather aspects which are
+used directly, but also aspects that are used indirectly, through virtual
+functions.
+
+For that the pass should compile a list of aspects used by each set of
+indirectly callable functions (as defined by `indirectly_callable` property set
+by user) and then append those aspects to every kernel which use those sets (as
+defined by `calls_indirectly` property set by user).
+
+**TODO**: should we consider outlining "indirectly used" aspects into a separate
+metadata and device image property? This should allow for more precise and
+user-friendly exceptions at runtime
+
+NOTE: if the aspects propagation pass is ever extended to track function
+pointers, then aspects attached to virtual functions **should not** be attached
+to kernels using this mechanism. For example, if a kernel uses a variable,
+which is initialized with a function pointer to a virtual function which uses
+an aspect, then such kernel **should not** be considered as using that aspect.
+Properties-based mechanism which is described above should be used for aspects
+propagation for virtual functions.
+
+To illustrate this, let's once again consider the example from Overview section
+which is copied below for convenience:
+
+```c++
+using syclext = sycl::ext::oneapi::experimental;
+
+struct set_fp64;
+
+struct Base {
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {}
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_fp64) void bar() {
+    // this virtual function uses double
+    double d = 3.14;
+  }
+};
+
+class Constructor;
+class Use;
+class UseFP64;
+
+int main() {
+  // Selected device may not support 'fp64' aspect
+  sycl::queue Q;
+
+  Base *Obj = sycl::malloc_device<Base>(1, Q);
+  int *Result = sycl::malloc_shared<int>(2, Q);
+
+  Q.single_task<Constructor>([=] {
+    // Even though at LLVM IR level this kernel does reference 'Base::foo'
+    // and 'Base::bar' though global variable containing `vtable` for `Base`,
```

Comment:
```suggestion
    // and 'Base::bar' through global variable containing `vtable` for `Base`,
```

Reply:
Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783